### ISSUE TITLE
test_remote_storage.py::test_remote_storage_backup_and_restore[local_fs]  hang on CI

### ISSUE THREAD

**Issue by hlinnaka** [neondatabase/neon#1626]
This CI run failed:

https://app.circleci.com/pipelines/github/neondatabase/neon/6017/workflows/9c2e39ba-1881-49ac-8a03-222309d33ef0/jobs/59320/steps

This was with PR 1616, but I'm pretty sure the code changes there did not cause this.

**Comment by hlinnaka** [comment]
There are interesting errors in the [pageserver log](https://output.circle-artifacts.com/output/job/6d0aae00-61ee-4993-a7e5-8c062ef1ef05/artifacts/0/tmp/test_output/test_remote_storage_backup_and_restore%5Blocal_fs%5D/repo/pageserver.log):
```
...
2022-05-03T22:32:47.593899Z  INFO storage_sync_loop_step:process_sync_tasks{sync_id=87b4f9506e884c108d3b0d3ecb8bdeec-32d71cf668c419457ffb0e30faff0434}: Processed a sync task in 4.10 seconds
2022-05-03T22:32:47.593911Z  INFO storage_sync_loop_step:process_sync_tasks{sync_id=87b4f9506e884c108d3b0d3ecb8bdeec-32d71cf668c419457ffb0e30faff0434}: Finished processing the task
2022-05-03T22:32:47.593937Z  INFO storage_sync_loop_step: Processing tasks for 1 timelines in batch, more tasks left to process: 0
2022-05-03T22:32:47.593950Z  INFO storage_sync_loop_step:process_sync_tasks{sync_id=87b4f9506e884c108d3b0d3ecb8bdeec-32d71cf668c419457ffb0e30faff0434}: Sync task 'download' processing started, attempt #4
2022-05-03T22:32:47.593957Z  INFO storage_sync_loop_step:process_sync_tasks{sync_id=87b4f9506e884c108d3b0d3ecb8bdeec-32d71cf668c419457ffb0e30faff0434}: Waiting 8 seconds before starting the 'download' task
2022-05-03T22:32:49.096730Z ERROR query handler for 'callmemaybe 87b4f9506e884c108d3b0d3ecb8bdeec 32d71cf668c419457ffb0e30faff0434 host=127.0.0.1 port=15009 options='-c ztimelineid=32d71cf668c419457ffb0e30faff0434 ztenantid=87b4f9506e884c108d3b0d3ecb8bdeec pageserver_connstr=postgresql://no_user:@localhost:15006'' failed: Cannot load local timeline
2022-05-03T22:32:50.129815Z  INFO Handling timeline 32d71cf668c419457ffb0e30faff0434 attach for tenant: 87b4f9506e884c108d3b0d3ecb8bdeec
2022-05-03T22:32:50.130022Z ERROR Error processing HTTP request: Conflict("Timeline download is already in progress")
2022-05-03T22:32:50.130046Z  INFO POST /v1/tenant/87b4f9506e884c108d3b0d3ecb8bdeec/timeline/32d71cf668c419457ffb0e30faff0434/attach 409 Conflict
2022-05-03T22:32:50.132774Z  INFO GET /v1/tenant/87b4f9506e884c108d3b0d3ecb8bdeec/timeline/32d71cf668c419457ffb0e30faff0434 200 OK
2022-05-03T22:32:50.141717Z  INFO Got SIGTERM. Terminating gracefully in fast shutdown mode
2022-05-03T22:32:50.141759Z  INFO waiting for libpq endpoint thread to shut down
2022-05-03T22:32:50.141846Z  INFO waiting for Remote storage sync thread to shut down
2022-05-03T22:32:50.141887Z  INFO Shutdown requested, stopping
2022-05-03T22:32:50.142358Z  INFO waiting for http_endpoint_thread to shut down
2022-05-03T22:32:50.142596Z  INFO Shut down successfully completed
2022-05-03T22:32:50.164721Z  INFO version: git:10f3df3f6271b7096d0a27264cfec0aa8719cdce
2022-05-03T22:32:50.164748Z  INFO Starting pageserver http handler on localhost:15008
2022-05-03T22:32:50.165086Z  INFO Starting pageserver pg protocol handler on localhost:15006
2022-05-03T22:32:50.165145Z  INFO daemonizing...
2022-05-03T22:32:50.165164Z ERROR could not daemonize err=unable to lock pid file
2022-05-03T22:32:50.165188Z  INFO Using auth: Trust
2022-05-03T22:32:50.165276Z  INFO removing temp download file at /tmp/test_output/test_remote_storage_backup_and_restore[local_fs]/repo/tenants/87b4f9506e884c108d3b0d3ecb8bdeec/timelines/32d71cf668c419457ffb0e30faff0434/000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__00000000016D0259-00000000016E3EF1.temp_download
2022-05-03T22:32:50.165395Z  INFO removing temp download file at /tmp/test_output/test_remote_storage_backup_and_restore[local_fs]/repo/tenants/87b4f9506e884c108d3b0d3ecb8bdeec/timelines/32d71cf668c419457ffb0e30faff0434/000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__00000000016965B0-0000000001696629.temp_download
2022-05-03T22:32:50.167458Z  INFO removing temp download file at /tmp/test_output/test_remote_storage_backup_and_restore[local_fs]/repo/tenants/87b4f9506e884c108d3b0d3ecb8bdeec/timelines/32d71cf668c419457ffb0e30faff0434/000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000001696629-00000000016D0259.temp_download
2022-05-03T22:32:50.167567Z ERROR Failed to process timeline dir contents at '/tmp/test_output/test_remote_storage_backup_and_restore[local_fs]/repo/tenants/87b4f9506e884c108d3b0d3ecb8bdeec/timelines/32d71cf668c419457ffb0e30faff0434', reason: No metadata file found in the timeline directory
...
```
The test shuts down the pageserver, deletes its data, and then restarts it. It kind of looks like the first pageserver didn't fully exit when the new one was started (`could not daemonize err=unable to lock pid file`). But then it started up anyway?

**Comment by hlinnaka** [comment]
This code in `pageserver.rs` looks suspicious:
```
        // XXX: The parent process should exit abruptly right after
        // it has spawned a child to prevent coverage machinery from
        // dumping stats into a `profraw` file now owned by the child.
        // Otherwise, the coverage data will be damaged.
        match daemonize.exit_action(|| exit_now(0)).start() {
            Ok(_) => info!("Success, daemonized"),
            Err(err) => error!(%err, "could not daemonize"),
        }
```
So, if the Daemonize::start() call fails, we print the error, but continue running. Surely we should exit on failure.

**Comment by hlinnaka** [comment]
The safekeeper startup code seems to have the same issue

**Comment by hlinnaka** [comment]
Fixed in commit 8346aa3a2
